cudax/samples/vector_add_async/vector_add.cu

```diff
@@ -0,0 +1,142 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Vector addition: C = A + B.
+ *
+ * This sample is a very basic sample that implements element by element
+ * vector addition. It is the same as the sample illustrating Chapter 2
+ * of the programming guide with some additions like error checking.
+ */
+
+#include <cuda/std/span>
+#include <cuda/std/type_traits>
+
+#include <cuda/experimental/launch.cuh>
+#include <cuda/experimental/stream.cuh>
+
+#include <cstdio>
+#include <vector>
+
+#include "async_copy.cuh"
+#include "async_memory_resource.cuh"
+#include "async_uninitialized_buffer.cuh"
+
+namespace cudax = cuda::experimental;
+
+/**
+ * CUDA Kernel Device code
+ *
+ * Computes the vector addition of A and B into C. The 3 vectors have the same
+ * number of elements numElements.
+ */
+__global__ void vectorAdd(cuda::std::span<const float> A, cuda::std::span<const float> B, cuda::std::span<float> C)
+{
+  int i = blockDim.x * blockIdx.x + threadIdx.x;
+
+  if (i < A.size())
+  {
+    C[i] = A[i] + B[i] + 0.0f;
+  }
+}
+
+/**
+ * Host main routine
+ */
+int main(void)
+try
+{
+  // A CUDA stream on which to execute the vector addition kernel
+  cudax::stream stream(cudax::devices[0]);
+
+  // Print the vector length to be used, and compute its size
+  int numElements = 50000;
+  std::printf("[Vector addition of %d elements]\n", numElements);
+
+  // Allocate the host vectors
+  std::vector<float> h_A(numElements); // input
+  std::vector<float> h_B(numElements); // input
+  std::vector<float> h_C(numElements); // output
+
+  // Initialize the host input vectors
+  for (int i = 0; i < numElements; ++i)
+  {
+    h_A[i] = rand() / (float) RAND_MAX;
+    h_B[i] = rand() / (float) RAND_MAX;
+  }
+
+  // This is a resource for allocating device memory
+  cudax::cuda_async_memory_resource mr(stream.device());
+
+  // Asynchronously allocate the device buffers on the stream
+  cudax::async_uninitialized_buffer d_A(float{}, mr, numElements, stream);
+  cudax::async_uninitialized_buffer d_B(float{}, mr, numElements, stream);
+  cudax::async_uninitialized_buffer d_C(float{}, mr, numElements, stream);
+
+  // Asynchronously copy the host input data to the device
+  cudax::async_copy(h_A, d_A, stream);
+  cudax::async_copy(h_B, d_B, stream);
+
+  // Define the kernel launch parameters
+  constexpr int threadsPerBlock = 256;
+  auto dims                     = cudax::distribute<threadsPerBlock>(numElements);
+
+  // Launch the vectorAdd kernel
+  std::printf(
+    "CUDA kernel launch with %d blocks of %d threads\n", dims.count(cudax::block, cudax::grid), threadsPerBlock);
+  cudax::launch(stream, dims, vectorAdd, d_A, d_B, d_C);
+
+  // Copy the result from device to host
+  cudax::async_copy(d_C, h_C, stream);
+
+  std::printf("waiting for the stream to finish\n");
+  stream.wait();
```

Comment:
Does successful return (without throwing) of `stream.wait()` indicate success?  That is, does this code demonstrate the error handling pattern that we want users to follow?

Reply:
Yes